ethcore/light/src/transaction_queue.rs

```diff
@@ -0,0 +1,421 @@
+	}
+
+	// attempt to move transactions from the future queue into the current queue.
+	fn adjust_future(&mut self) {
+		let mut next_nonce = self.next_nonce();
+
+		loop {
+			match self.future.remove(&next_nonce) {
+				Some(tx) => self.current.push(tx),
+				None => break,
+			}
+
+			next_nonce = next_nonce + 1.into();
+		}
+	}
+}
+
+/// Light transaction queue. See module docs for more details.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct TransactionQueue {
+	by_account: HashMap<Address, AccountTransactions>,
+	by_hash: H256FastMap<PendingTransaction>,
+}
+
+impl TransactionQueue {
+	/// Import a pending transaction to be queued.
+	pub fn import(&mut self, tx: PendingTransaction) -> Result<TransactionImportResult, TransactionError>  {
+		let sender = tx.sender();
+		let hash = tx.hash();
+		let nonce = tx.nonce;
+
+	    let res = match self.by_account.entry(sender) {
+			Entry::Vacant(entry) => {
+				entry.insert(AccountTransactions {
+					cur_nonce: CurrentNonce::Assumed(nonce),
+					current: vec![tx.clone()],
+					future: BTreeMap::new(),
+				});
+
+				TransactionImportResult::Current
+			}
+			Entry::Occupied(mut entry) => {
+				let acct_txs = entry.get_mut();
+				if &nonce < acct_txs.cur_nonce.value() {
+					// don't accept txs from before known current nonce.
+					if acct_txs.cur_nonce.is_known() {
+						return Err(TransactionError::Old)
+					}
+
+					// lower our assumption until corrected later.
+					acct_txs.cur_nonce = CurrentNonce::Assumed(nonce);
+				}
+
+				match acct_txs.current.binary_search_by(|x| x.nonce.cmp(&nonce)) {
+					Ok(idx) => {
+						trace!(target: "txqueue", "Replacing existing transaction from {} with nonce {}",
+							sender, nonce);
+
+						acct_txs.current[idx] = tx.clone();
+
+						TransactionImportResult::Current
+					}
+					Err(idx) => {
+						let cur_len = acct_txs.current.len();
+						let incr_nonce = nonce + 1.into();
+
+						// current is sorted with one tx per nonce,
+						// so if a tx with given nonce wasn't found that means it is either
+						// earlier in nonce than all other "current" transactions or later.
+						debug_assert!(idx == 0 || idx == cur_len);
+
+						if idx == 0 && acct_txs.current.first().map_or(false, |f| f.nonce != incr_nonce) {
+							let old_cur = ::std::mem::replace(&mut acct_txs.current, vec![tx.clone()]);
+
+							trace!(target: "txqueue", "Moving {} transactions with nonce > {} to future",
+								old_cur.len(), incr_nonce);
+
+							for future in old_cur {
+								let future_nonce = future.nonce;
+								acct_txs.future.insert(future_nonce, future);
+							}
+
+							TransactionImportResult::Current
+						} else if idx == cur_len && acct_txs.current.last().map_or(false, |f| f.nonce + 1.into() != nonce) {
+							trace!(target: "txqueue", "Queued future transaction for {}, nonce={}", sender, nonce);
+							let future_nonce = nonce;
+							acct_txs.future.insert(future_nonce, tx.clone());
```

Comment:
Future queue should have a hard limit (probably current queue also), otherwise you can use arbitrary amount of nodes memory just by propagating future transactions that will never be included in the blockchain.